### PR TITLE
(FACT-1574) add facter test for alternate fact directories

### PR DESCRIPTION
**Pull request by ericldelaney**
This adds a test for the 2 supported alternate directories where
facts can be found on Unix systems and tests fact precedence for them
TestRail: C59201

### CODE REVIEW

File: acceptance/lib/facter/acceptance/user_fact_utils.rb [hunk 1]
```diff
@@ -33,6 +33,26 @@ def get_factsd_dir(platform, version)
         end
       end
 
+      # Retrieve the path of the facts.d directory in /etc/facter on Unix systems
+      #
+      def get_etc_factsd_dir(platform, version)
```
Comment (Magisus) [inline acceptance/lib/facter/acceptance/user_fact_utils.rb]: If you're not using the version param here, might as well not pass it in...

File: acceptance/lib/facter/acceptance/user_fact_utils.rb [hunk 2]
```diff
@@ -33,6 +33,26 @@ def get_factsd_dir(platform, version)
...
+        if platform =~ /windows/
+          raise "get_etc_factsd_dir: not a supported directory on Windows"
+        else
+          File.join('/', 'etc', 'facter', 'facts.d')
+        end
+      end
+
+      # Retrieve the path of the facts.d diretory in /etc/puppetlabs/facter on Unix systems
+      #
+      def get_etc_puppetlabs_factsd_dir(platform, version)
```
Comment (Magisus) [inline acceptance/lib/facter/acceptance/user_fact_utils.rb]: Same here, unused version param can probably be removed.

File: acceptance/tests/external_facts/fact_directory_precedence.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  # ensure the fact directory we want to use exists
+  step "Agent #{agent}: create facts directory (#{etc_puppetlabs_factsd_dir})"
+  on(agent, "rm -rf #{etc_puppetlabs_factsd_dir}")
+  on(agent, "mkdir -p #{etc_puppetlabs_factsd_dir}")
+
+  # A fact in the etc_puppetlabs_factsd_dir directory should resolve to the fact
+  step "Agent #{agent}: create and resolve a custom fact in #{etc_puppetlabs_factsd_dir}"
+  create_remote_file(agent, etc_puppetlabs_factsd_path, ext_fact('etc_puppetlabs_path'))
+  on(agent, facter("test"))
```
Comment (Magisus) [inline acceptance/tests/external_facts/fact_directory_precedence.rb]: So in general, I have seen the following idiom to be more common:
```
on(agent, facter("test")) do
    assert_match(/etc_puppetlabs_path/, stdout, "Fact from #{etc_puppetlabs_factsd_dir} did not resolve correctly")
end
```

I think that is preferred so you can print a descriptive message of the failure. The functionality should be largely the same though, so might want to get a second opinion. The same is true for all the following test cases.

File: acceptance/tests/external_facts/fact_directory_precedence.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  assert_equal("ETC_FACTS", stdout.chomp)
+
+  # A fact in the etc_puppetlabs_factsd_dir should take precedence over the same fact in etc_factsd_dir
+  step "Agent #{agent}: create and resolve 2 facts of the same name between #{etc_factsd_dir} and #{etc_puppetlabs_factsd_dir}"
+  create_remote_file(agent, etc_factsd_path, ext_fact('ETC_FACTS'))
+  create_remote_file(agent, etc_puppetlabs_factsd_path, ext_fact('ETC_PUPPETLABS_FACTS'))
+  on(agent, facter("test"))
+  assert_equal("ETC_PUPPETLABS_FACTS", stdout.chomp)
+
+  teardown do
```
Comment (Magisus) [inline acceptance/tests/external_facts/fact_directory_precedence.rb]: I always used to forget this 👍

**Comment by Magisus** [comment]
Tried it out on a canary OS, passed, 👍  from me.

**Comment by LuvCurves** [comment]
👍 LGTM

**Pull request merged**